File: hotel-mern/controllers/type.controller.ts
import {
  basketService,
  buildingService,
  orderService,
  reviewService,
  roomService,
  typeService,
} from '../services'
import { ModifiedRequest } from '../shared/types'
import { Response } from 'express'
import { OrderPopulated, RoomPopulated, Type } from '../shared/models'
import { errorHandler } from '../shared/decorators'

export default class TypeController {
  async get(req: ModifiedRequest, res: Response) {
    const types: Type[] = await typeService.get()
    return res.json(types)
  }

  @errorHandler
  async create(req: ModifiedRequest, res: Response) {
    const { _services, name, places } = req.body
    const type: Type = await typeService.create(_services, name, places)
    return res.json(type)
  }

  @errorHandler
  async change(req: ModifiedRequest, res: Response) {
    const { _id, _services, name, places } = req.body
    const type: Type = await typeService.change(_id, _services, name, places)
    return res.json(type)
  }

  @errorHandler
  async delete(req: ModifiedRequest, res: Response) {
    const id: string = await typeService.delete(req.body._id)
    const rooms: RoomPopulated[] = await roomService.get({ _type: id })
    await roomService.deleteWithType(id)

    rooms.map(async (room: RoomPopulated) => {
      await reviewService.deleteWithRoom(room._id)
      await buildingService.removeRoom(room._id)

      if (room._order) {
        const order: OrderPopulated = await orderService.getOne(room._order)
        await basketService.removeOrder(order)
        await orderService.delete(room._order)
      }
    })

    return res.json(id)
  }
}

File: hotel/backend/src/main/kotlin/loshica/api/hotel/controllers/TypeController.kt
package loshica.api.hotel.controllers

import loshica.api.hotel.annotations.Auth
import loshica.api.hotel.dtos.DeleteDto
import loshica.api.hotel.dtos.TypeDto
import loshica.api.hotel.dtos.TypePopulatedDto
import loshica.api.hotel.models.Type
import loshica.api.hotel.interfaces.*
import loshica.api.hotel.models.User
import loshica.api.hotel.shared.Destroyer
import loshica.api.hotel.shared.Role
import loshica.api.hotel.shared.Route
import loshica.api.hotel.shared.Selector
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(Route.TYPES)
@CrossOrigin(originPatterns = ["*"])
class TypeController(
    bookingService: IBookingService,
    buildingService: IBuildingService,
    commentService: ICommentService,
    private val optionService: IOptionService,
    roomService: IRoomService,
    private val typeService: ITypeService,
    userService: IUserService
) {

    private val destroyer = Destroyer(
        bookingService = bookingService,
        buildingService = buildingService,
        commentService = commentService,
        optionService = optionService,
        roomService = roomService,
        typeService = typeService,
        userService = userService
    )

    @GetMapping
    fun getAll(): List<TypePopulatedDto> = typeService.getAll()
        .filter { it.isActive }
        .map { it.toPopulatedDto() }

    @GetMapping(Selector.ID)
    fun getOne(@PathVariable id: Int): TypePopulatedDto = typeService.getOne(id).toPopulatedDto()

    @PostMapping
    fun create(
        @Auth(Role.ADMIN) user: User,
        @RequestBody dto: TypeDto
    ): TypePopulatedDto {
        return typeService.create(
            name = dto.name,
            places = dto.places,
            price = dto.price,
            options = optionService.getByIds(dto.options)
        ).toPopulatedDto()
    }

    @PutMapping(Selector.ID)
    fun change(
        @Auth(Role.ADMIN) user: User,
        @RequestBody dto: TypeDto,
        @PathVariable id: Int
    ): TypePopulatedDto {
        return typeService.change(
            id = id,
            name = dto.name,
            places = dto.places,
            price = dto.price,
            options = optionService.getByIds(dto.options)
        ).toPopulatedDto()
    }

    @DeleteMapping(Selector.ID)
    fun delete(
        @Auth(Role.ADMIN) user: User,
        @PathVariable id: Int
    ): DeleteDto {
        val type: Type = typeService.getOne(id)
        type.rooms.forEach { destroyer.deleteRoom(it) }
        return DeleteDto(id = typeService.disable(id))
    }
}